solutions/gold/izho-17-LongestBeautifulSequence.mdx

```diff
@@ -0,0 +1,374 @@
+		- $\texttt{dp}[a_1].\texttt{end} = i = 1$.
+		- $\texttt{prv}[i] = \texttt{dp}[x].\texttt{end} = 0$.
+- $\texttt{dp}[a_2 = 3] = \{3, 2\}$.
+	- here, $x \in \{1, 2\}$ and $\texttt{dp}[x = 2]$ has the greater value, so:
+		- $\texttt{dp}[a_2]. \texttt{len} = \texttt{dp}[x].\texttt{len} + 1 = 3$.
+		- $\texttt{dp}[a_2].\texttt{end} = i = 2$.
+		- $\texttt{prv}[i] = \texttt{dp}[x].\texttt{end} = 1$.
+- $\texttt{dp}[a_3 = 4] = \{4, 3\}$.
+	- here, $x \in \{0, 1, 2, 3\}$ and $\texttt{dp}[x = 3]$ has the greatest value so:
+		- $\texttt{dp}[a_3]. \texttt{len} = \texttt{dp}[x].\texttt{len} + 1 = 4$.
+		- $\texttt{dp}[a_3].\texttt{end} = i = 3$.
+		- $\texttt{prv}[i] = \texttt{dp}[x].\texttt{end} = 2$.
+- **Final answer:** the max DP `len` value, $\boxed{4}$.
+
+### Implementation
+
+**Time Complexity**: $\mathcal{O}(NM)$
+
+<Spoiler title="Code">
+
+<LanguageSection>
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+using pii = pair<int, int>;
+const int B = 8;
+struct State {
+	int len, end;
+};
+int main() {
+	// preprocess:
+	// bc[i][j] = bit_count(i & j)
+	vector<vector<int>> bc(1 << B, vector<int>(1 << B));
+	for (int i = 0; i < (1 << B); i++) {
+		for (int j = 0; j < (1 << B); j++) {
+			bc[i][j] = __builtin_popcount(i & j);
+		}
+	}
+
+	int n;
+	cin >> n;
+	vector<int> a(n);
+	for (int i = 0; i < n; i++) cin >> a[i];
+	vector<int> k(n);
+	for (int i = 0; i < n; i++) cin >> k[i];
+	int ans = 1;
+	// best_i: most optimal index to end on
+	int best_i = 0;
+	vector<int> prv(n);               // prv[i]: best index to include before i
+	iota(prv.begin(), prv.end(), 0);  // prv[i] = i
+	vector<State> dp(1 << B, State{-1, -1});
+
+	for (int i = 0; i < n; i++) {
+		if (dp[a[i]].len == -1) {
+			dp[a[i]].len = 1;
+			dp[a[i]].end = i;
+		}
+		for (int j = 0; j < (1 << B); j++) {
+			if (bc[j][a[i]] == k[i] && dp[j].len + 1 > dp[a[i]].len) {
+				dp[a[i]].len = dp[j].len + 1;
+				dp[a[i]].end = i;
+				prv[i] = dp[j].end;
+			}
+		}
+		if (dp[a[i]].len > ans) {
+			ans = dp[a[i]].len;
+			best_i = i;
+		}
+	}
+	cout << ans << endl;
+	vector<int> res;
+	while (prv[best_i] != best_i) {
+		res.push_back(best_i);
+		best_i = prv[best_i];
+	}
+	res.push_back(best_i);
+	reverse(res.begin(), res.end());
+	for (int x : res) cout << x + 1 << " ";
+}
+```
+
+</CPPSection>
+
+</LanguageSection>
+</Spoiler>
+
+### Subtask 4 (Full Credit)
+
+With much higher bounds on both $N$ and $M$, neither of the previously presented solutions will run in time.
+
+In our solution for Subtask 3, the loop from $0...M$ seems repetitive and motivates somehow restricting the search space of DP states. **How can we quickly process all $x$ in $\texttt{trans}_{k_i}(a_i)$?**
+
+The most extreme way to optimize transitions is to embed the restrictions of $x$ directly in the DP state.
+
+**Concretely, define $\texttt{dp2}[i][k]$ as the maximum value of $\texttt{dp}[x]$ over all $x$ in $\texttt{trans}_k(i)$.**
```

Comment:
i'd like it if these variables were more self describing, but up to you at this point

Reply:
i think the current variable names are more concise but ill take any suggestions into consideration

Reply:
you define this dp2 but i never see it in your sol?

Reply:
dp2 is more of a transition toward dp3 than an actual solution
if you think i would be helpful, I could also provide a solution to subtask 3 using dp2 rather than dp

Reply:
think of it like uh
"oh, there's this variable, wonder how they actually use it in the code"
"...where is it"
ctrl+f "dp2"
"WHERE IS IT WHAT"

Reply:
haha when you put it like that 😅